proposals/0176-static-syscalls.md

```diff
@@ -0,0 +1,87 @@
+---
+simd: '0178'
+title: SBPF Static Syscalls
+authors:
+  - Alessandro Decina
+  - Alexander Meißner
+  - Lucas Steuernagel
+category: Standard
+type: Core
+status: Draft
+created: 2024-09-27
+---
+
+## Summary
+
+This SIMD introduces a new instruction syscall in the SBPF instruction set to 
+represent syscalls. Such a change aims to remove relocations when resolving 
+syscalls and simplify the instruction set, allowing for the straightforward 
+differentiation between external and internal calls.
+
+## Motivation
+
+The resolution of syscalls during ELF loading requires relocating addresses, 
+which is a performance burden for the validator. Relocations require an entire 
+copy of the ELF file in memory to either relocate addresses we fetch from the 
+symbol table or offset addresses to after the start of the virtual machine’s 
+memory. Moreover, relocations pose security concerns, as they allow the 
+arbitrary modification of program headers and programs sections. A new 
+separate opcode for syscalls modifies the behavior of the ELF loader, allowing 
+us to resolve syscalls without relocations.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+The following must go into effect if and only if a program indicates the SBPF 
+version XX or higher in its ELF header e_flags field, according to the 
+specification of SIMD-0161.
+
+### New syscall instruction
+
+We introduce a new instruction in the SBPF instruction set, which we call 
+`syscall`. It must be associated with all syscalls in the SBPF format. Its 
+encoding consists of an opcode `0x95` and an immediate, which must refer to a 
+previously registered syscall. For more reference on the SBF ISA format, see 
+the 
+[spec document](https://github.com/solana-labs/rbpf/blob/main/doc/bytecode.md).
+
+For simplicity, syscalls must be represented as a natural number greater than 
+zero, so that they can be organized in a lookup table. This choice allows for 
+quick retrieval of syscall information from integer indexes. An instruction 
+`syscall 2` must represent a call to the function registered at position two 
+in the lookup table.
+
+Consequently, system calls in the Solana SDK and in any related compiler tools 
+must be registered as function pointers, whose address is a natural number 
+greater than zero, representing their position in a syscall lookup table. The 
```

Comment:
Sorry if I'm missing the extra context here, but is this syscall lookup table part of this SIMD proposal, or another?

Reply:
It is part of this one. The idea is to remove the hash map we use for lookups today. There was a request to use integer indexing for syscalls to avoid hashes. The underlying data structure does not impact the protocol and is more implementation specific. Did my wording sound too vague?

Reply:
I think the wording is totally fine, I just wonder if we should have a table of all of those integer IDs in this proposal itself, since this would definitely need to be matched on all clients. What do you think?

Reply:
Yes, I'll add the table.

Reply:
Added.

Reply:
Thanks! Looks great. Any particular ordering, or just arbitrarily selected?

Reply:
> Any particular ordering, or just arbitrarily selected?

I tried grouping related syscalls to similar numbers and I wanted those declared in the compiler to have the lower numbers.